ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -487,7 +483,7 @@ Operator genOPTree(PlannerContext plannerCtx) throws SemanticException {
             ASTNode newAST = getOptimizedAST(newPlan);
 
             // 1.1. Fix up the query for insert/ctas/materialized views
-            newAST = fixUpAfterCbo(this.getAST(), newAST, cboCtx);
```

Comment:
I don't see how this change will not reintroduce the issue fixed in HIVE-22578

because the "fixUpAfterCbo" makes calls to a function named replaceASTChild which changes the actual ast - and it may make it impossible to fallback to the non-cbo path

Reply:
@kgyrtkirk The original issue for which HIVE-22578 was opened is being fixed by HIVE-22824 (this pull request's change). CBO path was failing because JoinProjectTranspose rule was removing project containing windowing (creating wrong AST).
Fall to non-cbo path should happen only for queries for which CBO isn't supported (and that will happed before fixUpAfterCbo). So I believe it is okay to change AST at this point.

Reply:
ok; after all we want to have cbo on more often than off